spec.bs

```diff
@@ -324,27 +327,32 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
-    1. If |window|'s [=associated document=] is not [=fully active=], return a [=promise rejected=] with a {{TypeError}}.
+    1. If |window|'s [=Window/browsing context=] is null, then return the [=tuple=] of a ([=promise rejected=] with a {{TypeError}}, true).
+    1. If |window|'s [=associated document=] is not [=fully active=], return the [=tuple=] of a ([=promise rejected=] with a {{TypeError}}, true).
     1. [=Assert=]: |worklet|'s [=global scopes=]'s [=list/size=] is 1.
     1. Let |globalScope| be |worklet|'s [=global scopes=][0].
     1. Let |moduleMapKeyTuples| be the result of running [=map/get the keys=] on |globalScope|'s [=relevant settings object=]'s [=module map=].
     1. [=Assert=]: |moduleMapKeyTuples| has [=map/size=] 1.
     1. Let |moduleURLRecord| be |moduleMapKeyTuples|[0][0].
     1. Let |savedQueryName| be |options|["`savedQuery`"].
     1. If |savedQueryName| is a [=string=] that is not the empty string, then:
-        1. Let |savedIndex| be the result of running [=get the index for a saved query=] on |navigable|, |workletDataOrigin|, |moduleURLRecord|, |operationName|, and |savedQueryName|.
-        1. If |savedIndex| is not null, then:
-            1. If |savedIndex| is greater than |urlList|'s [=list/size=], return a [=promise rejected=] with a {{TypeError}}.
-
-                Note: The result index is beyond the input urls' size. This violates the selectURL() protocol, and we don't know which url should be selected.
-            1. Return a [=promise resolved=] with |savedIndex|.
-    1. Return |promise|, and immediately [=obtain a worklet agent=] given |window| and run the rest of these steps in that agent:
+        1. Let |callbackTask| be the result of running [=obtain a callback to process the saved index result=], given |window|, |urlList|, and |promise|.
+        1. Let |savedIndex| be the result of running [=get the index for a saved query=] on |navigable|, |workletDataOrigin|, |moduleURLRecord|, |operationName|, |savedQueryName|, and |callbackTask|.
+        1. If |savedIndex| is "pending callback", then return the [=tuple=] (|promise|, false).
+
+            Note: |callbackTask| is now stored to be run when a previously obtained worklet agent completes its operation to select the index for this query. When the steps of |callbackTask| are run, |promise| will be resolved.
+        1. If |savedIndex| is an {{unsigned long}}, then:
+            1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to run the steps of |callbackTask|, given |savedIndex|.
+
+                Note: Running the steps of |callbackTask| will resolve |promise|.
+            1. Return the [=tuple=] (|promise|, false).
+        1. [=Assert=] that |savedIndex| is "obtain worklet agent".
+    1. Return the [=tuple=] (|promise|, true), and immediately [=obtain a worklet agent=] given |window| and run the rest of these steps in that agent:
```

Comment:
I haven't seen this "return foo and then do something async" style before.  I think it would be preferred to structure it like this:

```
1. [=Obtain a worklet agent=] given |window| and run the following steps in that agent:
  1. // fill in async steps
1. Return the [=tuple=] (|promise|, true).
```

The async steps will effectively continue to run after the return here.

Reply:
Although I think we should `[=queue a task=]` on the worklet agent's event loop instead of using the vague "run the steps in that agent".

See:

https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task
https://html.spec.whatwg.org/#concept-agent-event-loop